### PR TITLE
SuperGLUE Diagnostics

### PR DESCRIPTION
**Pull request by W4ngatang**
Implement the SuperGLUE diagnostic task. Mostly re-use mnli-diagnostic but generalize it into a new class, GLUEDiagnosticTask.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -394,11 +395,11 @@ def main(cl_arguments):
                 "Error: ELMo scalars loaded and will be updated in do_target_task_training but "
                 "they should not be updated! Check sep_embs_for_skip flag or make an issue.")
         for task in target_tasks:
-            # Skip mnli-diagnostic
+            # Skip diagnostic tasks b/c they should not be trained on
             # This has to be handled differently than probing tasks because probing tasks require the "is_probing_task"
             # to be set to True. For mnli-diagnostic this flag will be False because it is part of GLUE and
             # "is_probing_task is global flag specific to a run, not to a task.
-            if task.name == 'mnli-diagnostic':
+            if isinstance(task, GLUEDiagnosticTask):
```
Comment (sleepinyourhat) [inline main.py]: What about MultiNLIDiagnosticTask?

File: main.py [hunk 2]
```diff
@@ -469,7 +470,7 @@ def main(cl_arguments):
         splits_to_write = evaluate.parse_write_preds_arg(args.write_preds)
         if args.transfer_paradigm == "finetune":
             for task in target_tasks:
-                if task.name == 'mnli-diagnostic':
+                if isinstance(task, GLUEDiagnosticTask):
```
Comment (sleepinyourhat) [inline main.py]: ''

File: src/tasks/tasks.py [hunk 3]
```diff
@@ -1247,6 +1247,211 @@ def collect_metrics(ix_to_tag_dict, tag_group):
         collect_metrics(self.ix_to_knowledge_dic, 'knowledge')
         return collected_metrics
 
+@register_task('mnli-diagnostic', rel_path='MNLI/', n_classes=3)
```
Comment (sleepinyourhat) [inline src/tasks/tasks.py]: But there's still code for MNLIDiagnosticTask elsewhere...

File: src/evaluate.py [hunk 4]
```diff
@@ -119,7 +119,7 @@ def evaluate(
         all_preds[task.name] = task_preds
         log.info("Finished evaluating on: %s", task.name)
 
-    all_metrics["micro_avg"] /= n_examples_overall
+    all_metrics["micro_avg"] /= max(n_examples_overall, 1) # hack for diagnostics
```
Comment (sleepinyourhat) [inline src/evaluate.py]: Whitespace?

File: src/evaluate.py [hunk 5]
```diff
@@ -168,6 +168,12 @@ def write_preds(
             _write_wic_preds(
                 task, preds_df, pred_dir, split_name, strict_glue_format=strict_glue_format
             )
+
```
Comment (sleepinyourhat) [inline src/evaluate.py]: Delete line.

File: src/evaluate.py [hunk 6]
```diff
@@ -168,6 +168,12 @@ def write_preds(
...
+        elif isinstance(task, GLUEDiagnosticTask):
+            # glue-diagnostic is caught above by being in ALL_GLUE_TASKS
```
Comment (sleepinyourhat) [inline src/evaluate.py]: Wait, so what does this elif case catch? Just superglue?

Comment (W4ngatang) [reply src/evaluate.py]: Currently yea

**Pull request merged**